src/client/index.ts

```diff
@@ -18,111 +79,268 @@ export interface NotificationOptions {
+     *
+     * If omitted or `null`, notification will persist until user explicitly actions/clears it.
      */
-    buttons?: OptionButton[];
-}
+    expires?: Date|null;
 
-/**
- * Configuration options for constructing a Button within a Notification.
- */
-export interface OptionButton {
-    title: string;
-    iconUrl?: string;
+    /**
+     * A list of button definitions, to display below the notification text.
+     *
+     * Notifications support up to four buttons, attempting to add more than four will result in the notification
+     * being rejected by the service.
+     */
+    buttons?: ButtonOptions[];
+
+    /**
+     * An action that will fire when the notification is clicked. This only fires on clicks to the notification body,
+     * interactions with buttons (both application-defined buttons, and the default 'X' close button) will not trigger
+     * a select action.
+     *
+     * If omitted or `null`, applications will not receive a {@link NotificationActionEvent|`notification-action`}
+     * event when the notification is clicked. See {@link Actions} for more details on Notification actions, and
+     * receiving interaction events from notifications.
+     */
+    onSelect?: ActionDeclaration<never, never>|null;
 }
 
 /**
- * User-defined context data that can be attached to Notifications.
+ * Application-defined context data that can be attached to notifications.
  */
 export type CustomData = any;
 
 /**
- * A fully hydrated form of the {@link NotificationOptions}.
+ * A fully-hydrated form of {@link NotificationOptions}.
+ *
+ * After {@link create|creating} a notification, the service will return this object - which will be the given options
+ * object, with any unspecified fields filled-in with default values.
+ *
+ * This object should be treated as immutable, modifying its state will not have any effect on the notification that
+ * the user sees on-screen.
  */
-export type Notification = Required<NotificationOptions>;
+export type Notification = Readonly<NotificationOptions & Required<NotificationOptions>>;
 
 /**
- * Event fired whenever the Notification has been clicked on.
+ * Event that is fired for interactions with notification UI elements. It is important to note that applications will
+ * only receive these events if they indicate to the service that they want to receive these events. See
+ * {@link Actions} for a full example of how actions are defined, and how an application can listen to and handle them.
+ *
+ * This can be for the notification button(s). Later versions of the service will add additional control types. All
+ * actions, for all control types, will be returned to the application via the same `notification-action` event type.
+ * Check the contents of the event object for details on what triggered this action, which control the user
+ * interacted with, and what metadata was attached to the action.
```

Comment:
Feel iffy that we're using 'metadata' to describe the action result in this file, when we're not using that term elsewhere

Reply:
Tweaked the wording a bit, but still using 'metadata' here.

We do use it in the `NotificationActionResult` definition too (which is basically what we're talking about here), but wording could likely still be improved.

Reply:
Drop "to fetch"

Reply:
Missing and "and"